vnext/Mso/src/dispatchQueue/looperScheduler.cpp

```diff
@@ -50,25 +50,29 @@ LooperScheduler::~LooperScheduler() noexcept {
           if (auto &func = self->m_settings.TaskStarting) {
             func(q);
           }
-
           queue->InvokeTask(std::move(task), std::nullopt);
-
           if (auto &func = self->m_settings.TaskCompleted) {
             func(q);
           }
         }
+      }
 
-        if (self->m_isShutdown) {
-          break;
-        }
+      if (self->m_isShutdown) {
+        break;
+      }
 
+      if (auto queue = self->m_queue.GetStrongPtr()) {
```

Comment:
>GetStrongPtr [](http://example.com/codeflow?start=37&length=12)

Getting strong ptr and then creating a copy of CntPtr causes two interlocked operations which are better to avoid if we can. In this code it is better to do it only if there is a callback defined:

```
if (auto &func = self->m_settings.IdleWaitStarting) {
  if (auto queue = self->m_queue.GetStrongPtr()) {
     func(DispatchQueue{std::move(queue)});
  }
}
```

The std::move above helps to avoid unnecessary interlocked increment. Thus, we will have only one interlocked operation, and only if we have a callback.

The same is for the code below.